File: kevoree-platform/org.kevoree.platform.agent/src/main/java/org/kevoree/tools/agent/KevoreeNodeRunner.java
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kevoree.tools.agent;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;

public class KevoreeNodeRunner {

	private Logger logger = LoggerFactory.getLogger(KevoreeNodeRunner.class);

	private Process nodePlatformProcess;
	private String platformJARPath;

	private String nodeName;
	private Integer basePort;

	public KevoreeNodeRunner(String nodeName, Integer basePort) {

		//URL url = new URL(null,"cvs://server/project/folder#version", new PaxMvnUrlStreamHandlerFactory());
		//System.setProperty("org.ops4j.pax.url.mvn.defaultRepositories", "http://maven.kevoree.org/release");
		//URL.setURLStreamHandlerFactory(new PaxMvnUrlStreamHandlerFactory()); // to use the maven URL handler
		this.nodeName = nodeName;
		this.basePort = basePort;

	}

	public void startNode() {
		try {
			System.out.println("StartNodeCommand");
			if (platformJARPath == null) {
				getJar();
			}
			nodePlatformProcess = Runtime.getRuntime().exec(new String[]{"java", "-Dnode.name=" + nodeName, "-Dnode.port=" + basePort, "-jar", platformJARPath});
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}

	}

	public void stopKillNode() {
		System.out.println("KillNodeCommand");
		try {
			nodePlatformProcess.getOutputStream().write("stop 0".getBytes());
			nodePlatformProcess.getOutputStream().flush();
		} catch (IOException e) {
			//e.printStackTrace();
			logger.error("The node cannot be killed. Try to force kill", e.getCause());
			nodePlatformProcess.destroy();
		}
	}

	/*private void getJar() { // from maven(doesn't work ...)
		try {
			URL mvnURL = new URL("mvn:http://maven.kevoree.org/release/!org.kevoree.platform/org.kevoree.platform.osgi.standalone");
			InputStream stream = mvnURL.openConnection().getInputStream();

			File f = File.createTempFile("org.kevoree.platform.osgi.standalone", ".jar");
			f.deleteOnExit();
			OutputStream outputStream = new FileOutputStream(f);

			byte[] bytes = new byte[1024];
			int length = 0;
			while ((length = stream.read()) != -1) {
				outputStream.write(bytes, 0, length);
			}
			outputStream.flush();
			outputStream.close();
			stream.close();

			platformJARPath = f.getAbsolutePath();
		} catch (MalformedURLException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}*/

	/*private void getJar() { // from the availables resources into the jar
		try {
			InputStream stream = this.getClass().getClassLoader().getResourceAsStream("org.kevoree.platform.osgi.standalone.jar");

			File f = new File(defaultJarFilePath);
			f.deleteOnExit();
			OutputStream outputStream = new FileOutputStream(f);

			byte[] bytes = new byte[1024];
			int length;
			while ((length = stream.read()) != -1) {
				outputStream.write(bytes, 0, length);
			}
			outputStream.flush();
			outputStream.close();
			stream.close();

			platformJARPath = f.getAbsolutePath();
		} catch (MalformedURLException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}*/

	private void getJar() throws IOException {
		String jarLocation = System.getProperty("kevoree.location");
		if (jarLocation == null) {
			jarLocation = System.getProperty("user.dir") + "org.kevoree.platform.osgi.standalone" + getVersion() + ".jar";
		}
		if (new File(jarLocation).exists()) {
			platformJARPath = jarLocation;
		} else {
			throw new FileNotFoundException(jarLocation + " doesn't exist");
		}
	}

	private String getVersion() throws IOException {
		InputStream stream = this.getClass().getClassLoader().getResourceAsStream("META-INF/maven/org.kevoree.platform/org.kevoree.platform.agent/pom.properties");
		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

		byte[] bytes = new byte[1024];
		int length;
		while ((length = stream.read()) != -1) {
			outputStream.write(bytes, 0, length);
		}
		outputStream.flush();
		outputStream.close();
		stream.close();

		StringBuilder builder = new StringBuilder(outputStream.toString());
		int index = builder.indexOf("version=");
		builder.delete(0, index);
		int end = builder.indexOf(System.getProperty("line.separator"));
		builder.delete(end, builder.length());
		int equalsSign = builder.indexOf("=");
		builder.delete(0, equalsSign + 1);

		return builder.toString();
	}

	/*public class PaxMvnUrlStreamHandlerFactory implements URLStreamHandlerFactory {

		Handler handler;

		public PaxMvnUrlStreamHandlerFactory() {
			handler = new Handler();
		}

		@Override
		public URLStreamHandler createURLStreamHandler(String protocol) {
			return handler;
		}
	}*/
}

File: kevoree-library/sky/org.kevoree.library.sky.api/src/main/scala/org/kevoree/library/sky/manager/KevoreeNodeRunner.scala
package org.kevoree.library.sky.manager

import org.kevoree.framework.KevoreePropertyHelper
import org.slf4j.{LoggerFactory, Logger}
import org.kevoree.{TypeDefinition, ContainerRoot}
import java.io._

/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 20/09/11
 * Time: 11:46
 *
 * @author Erwan Daubert
 * @version 1.0
 */
abstract class KevoreeNodeRunner (var nodeName: String) {

  private val logger: Logger = LoggerFactory.getLogger(classOf[KevoreeNodeRunner])

  def startNode (iaasModel : ContainerRoot,jailBootStrapModel : ContainerRoot): Boolean

  def stopNode (): Boolean

  //def updateNode (modelPath: String): Boolean


  var outFile: File = null

  def getOutFile = outFile

  var errFile: File = null

  def getErrFile = errFile

  /**
   * configure the ssh server
   * @param model
   * @param path
   * @param ip
   */
  def configureSSHServer (model: ContainerRoot, path: String, ip: String) {
    if (ip != null && ip != "") {
      logger.debug("configure ssh server ip")
      replaceStringIntoFile("<ip_address>", ip, path + File.separator + "etc" + File.separator + "ssh" + File.separator + "sshd_config")
    }
  }

  private def isASubType (nodeType: TypeDefinition, typeName: String): Boolean = {
    nodeType.getSuperTypes.find(td => td.getName == typeName || isASubType(td, typeName)) match {
      case None => false
      case Some(typeDefinition) => true
    }
  }

  private def copyStringToFile (data: String, outputFile: String) {
    if (data != null && data != "") {
//      logger.debug("trying to copy \"{}\" to {}", data, outputFile)
      try {
        if (new File(outputFile).exists()) {
          new File(outputFile).delete()
        }
        val writer = new DataOutputStream(new FileOutputStream(new File(outputFile)))

        writer.write(data.getBytes)
        writer.flush()
        writer.close()
//        logger.debug("copying \"{}\" into {} uis done", data, outputFile)
      } catch {
        case _@e => logger.error("Unable to copy \"{}\" on {}", Array[AnyRef](data, outputFile), e)
      }
    }
  }

  private def replaceStringIntoFile (dataToReplace: String, newData: String, file: String) {
    if (dataToReplace != null && dataToReplace != "" && newData != null && newData != "") {
//      logger.debug("trying to replace \"{}\" by \"{}\" into {}", Array[AnyRef](dataToReplace, newData, file))
      if (new File(file).exists()) {
        try {
          val stringBuilder = new StringBuilder
          val reader = new DataInputStream(new FileInputStream(new File(file)))
          val writer = new ByteArrayOutputStream()

          val bytes = new Array[Byte](2048)
          var length = reader.read(bytes)
          while (length != -1) {
            writer.write(bytes, 0, length)
            length = reader.read(bytes)

          }
          writer.flush()
          writer.close()
          reader.close()
          stringBuilder append new String(writer.toByteArray)
          stringBuilder.replace(stringBuilder.indexOf(dataToReplace), stringBuilder.indexOf(dataToReplace) + dataToReplace.length(), newData)

          copyStringToFile(stringBuilder.toString(), file)
//          logger.debug("replacing \"{}\" by \"{}\" into {} is done", Array[AnyRef](dataToReplace, newData, file))
        } catch {
          case _@e => logger.error("Unable to replace a string", e)
        }
      } else {
        logger.debug("The file {} doesn't exist, nothing can be replace.", file)
      }
    }
  }
}

